### PR TITLE
[MRG] Add fine-tune CV model API

### PR DESCRIPTION
**Pull request by YuanmingLeee**
#### What has been done to verify that this works as intended?

First, make sure you have installed the correct **experimental** environment according to `CONTRIBUTE.md`. You have also installed the `modelci` pip package, to enable the CLI tools. You have also started the ModelCI service by:
```
source scripts/setup_env.sh
modelci start
```

Then register a `ResNet50, PyTorch` model into the ModelDB.
```
modelci model export -n ResNet50 -f PyTorch
```

Next, start the frontend. (Kill it first if you have started it before)
```
python -m uvicorn modelci.app.main:app --host 0.0.0.0 --reload
```

Now, go the the 0.0.0.0:8000/docs, try the API 
```
[PATCH] ​/api​/exp​/cv-tuner​/finetune​/{id}​
```
Click`Try it out`, find the model ID of the PyTorch model*, and click execute as follows:
image

#### Why is this the best possible solution? Were any other approaches considered?

To be discussed.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

A new feature, no breaking changes to the existing system.

#### Does this change require updates to documentation?

Yes.

#### Before submitting this PR, please make sure you have:

- [x] run `python -m pytest tests/` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited.

### CODE REVIEW

File: modelci/app/v1/endpoints/model.py [hunk 1]
```diff
@@ -78,11 +87,11 @@ def update_model_structure_as_new(id: str, structure, dry_run: bool = False):  #
             [A] fc1 -> fc2
 
         >>>
-        ... structure = {
+        ... structure_data = {
         ...     'layer': {
         ...         'fc': {'_op': 'D'},
```
Comment (univerone) [inline modelci/app/v1/endpoints/model.py:92]: `_op` should be change to `op_` in this example

File: modelci/app/v1/endpoints/model.py [hunk 2]
```diff
@@ -102,28 +111,121 @@ def update_model_structure_as_new(id: str, structure, dry_run: bool = False):  #
...
+    output_tensors = net(*input_tensors)
+    if not isinstance(output_tensors, (list, tuple)):
+        output_tensors = (output_tensors,)
+    for output_tensor in output_tensors:
+        output_shape = IOShape(shape=output_tensor.shape, dtype=type_to_data_type(output_tensor.dtype))
+        output_shapes.append(output_shape)
+
+    if not dry_run:
+        register_model(
+            net, dataset='', metric=model.metric, task=model.task,
```
Comment (univerone) [inline modelci/app/v1/endpoints/model.py:224]: Should we set model.metric to empty as the dataset ? As it will be updated during later on fintune process

**Review by univerone — DISMISSED** [review]
LGTM! This is a breakthrough new feature!

File: modelci/app/v1/endpoints/model.py (hunk 2 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py:224]: Yes, we should.

File: modelci/app/v1/endpoints/model.py [hunk 3]
```diff
@@ -49,20 +55,23 @@ async def get_model_structure(id: str):  # noqa
...
-def update_model_structure_as_new(id: str, structure, dry_run: bool = False):  # noqa
+def update_model_structure_as_new(id: str, structure: Structure, dry_run: bool = False):  # noqa
     """
     TODO: Update model structure by adjusting layers (add, modify, delete) or rewiring the
         connections between layers.
 
     Examples:
         Fine-tune the model by modify the layer with name 'fc' (last layer). The layer
-        has a changed argument out_features = 10. _op='M' indicates the operation to this layer ('fc')
+        has a changed argument out_features = 10. op_='M' indicates the operation to this layer ('fc')
```
Comment (HuaizhengZhang) [inline modelci/app/v1/endpoints/model.py:65]: op_这几个参数A，M之类的没必要用enum
增加了代码阅读难度

File: modelci/app/v1/endpoints/model.py [hunk 4]
```diff
@@ -78,11 +87,11 @@ def update_model_structure_as_new(id: str, structure, dry_run: bool = False):  #
...
+        ... structure_data = {
         ...     'layer': {
         ...         'fc': {'_op': 'D'},
-        ...         'fc1': {'in_features': 1024, 'out_features': 512, '_type': 'nn.Linear', '_op': 'A'},
-        ...         'fc2': {'in_features': 512, 'out_features': 10, '_type': 'nn.Linear', '_op': 'A'},
+        ...         'fc1': {'in_features': 1024, 'out_features': 512, 'type_': 'torch.nn.Linear', '_op': 'A'},
+        ...         'fc2': {'in_features': 512, 'out_features': 10, 'type_': 'torch.nn.Linear', '_op': 'A'},
         ...     },
         ...     'connection': {
         ...         'conv1': {'fc': 'D', 'fc1': 'A'},
```
Comment (HuaizhengZhang) [inline modelci/app/v1/endpoints/model.py:97]: 增加了两层，为什么这个connection只增加了一层fc1?

File: modelci/app/v1/endpoints/model.py (hunk 3 above)
Comment (HuaizhengZhang) [reply modelci/app/v1/endpoints/model.py:65]: 我建议直接全拼

File: modelci/app/v1/endpoints/model.py [hunk 5]
```diff
@@ -49,20 +55,23 @@ async def get_model_structure(id: str):  # noqa
 
 
 @router.patch('/structure/{id}')  # TODO: add response_model
-def update_model_structure_as_new(id: str, structure, dry_run: bool = False):  # noqa
+def update_model_structure_as_new(id: str, structure: Structure, dry_run: bool = False):  # noqa
```
Comment (HuaizhengZhang) [inline modelci/app/v1/endpoints/model.py:58]: 合并的时候，注意这个structure参数名换成updated_layer

File: modelci/app/v1/endpoints/model.py [hunk 6]
```diff
@@ -102,28 +111,121 @@ def update_model_structure_as_new(id: str, structure, dry_run: bool = False):  #
 
 
 @router.patch('/structure/{id}/finetune')
-def update_finetune_model_as_new(id: str, updated_layer: dict, dry_run: bool = False):  # noqa
+def update_finetune_model_as_new(id: str, updated_layer: Structure, dry_run: bool = False):  # noqa
```
Comment (HuaizhengZhang) [inline modelci/app/v1/endpoints/model.py:114]: 看到了comments明白了，good

File: modelci/hub/manager.py [hunk 7]
```diff
@@ -91,7 +92,21 @@ def register_model(
         else:  # from implicit extracted from path, check validity of the path later at registration
             path = model_dir
         model_dir_list.append(path)
-    else:
+    elif framework == Framework.PYTORCH and engine == Engine.PYTORCH:
```
Comment (HuaizhengZhang) [inline modelci/hub/manager.py:95]: 这块为啥不是shanshan完成的 @univerone

File: modelci/types/vo/model_structure_vo.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/env python
```
Comment (HuaizhengZhang) [inline modelci/types/vo/model_structure_vo.py:1]: 这种vo的jave写法逐渐替换掉吧，不要在这样写了，还是写的简单pythonic一些

File: modelci/app/v1/endpoints/model.py (hunk 4 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py:97]: You miss a connection:
```
'connection': {
         'conv1': {'fc': 'D', 'fc1': 'A'},
         'fc1': {'fc2': 'A'},
}
```
Delete `conv1 -> fc`, add `conv1 -> fc1` and `fc1 -> fc2`

File: modelci/app/v1/endpoints/model.py (hunk 3 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py:65]: By using Enum, we can leave the typing validation jobs to `pydantic`. If we use a string, we should check if an invalid string is input every time.

File: modelci/hub/manager.py (hunk 7 above)
Comment (YuanmingLeee) [reply modelci/hub/manager.py:95]: I change the logic a little bit. This function has a too complex logic, I will leave a TODO here.

Comment (YuanmingLeee) [reply modelci/hub/manager.py:95]: This function should be split.

File: modelci/types/vo/model_structure_vo.py (hunk 8 above)
Comment (YuanmingLeee) [reply modelci/types/vo/model_structure_vo.py:1]: Though the name is `..._vo`，it is fully written by `pydantic`, and can directly save to the database. 
The `types` package will be refactored in a Django-like layout in a few releases:
`types.bo` delete
`types.do` -> `types.orm`
`types.vo` -> `types.model`
others -> `types.common`

File: modelci/app/v1/endpoints/model.py (hunk 5 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py:58]: The `structure` parameter contains `layer` and `connections`, so that the data structure is the same as `update_structure`.
This is for easy migration of the frontend: just need to change the called url.

File: modelci/app/v1/endpoints/model.py (hunk 1 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py]: Updated

**Comment by HuaizhengZhang** [comment]
lgtm. You can creat a exp/ dir and put some code into it to speed up your progress. We will not review code in it.

**Review by HuaizhengZhang — APPROVED** [review]
